File: oop_phonebook/src/repl.rs
// pub mod repl;
pub use crate::*;

use std::io::Write;

pub enum Input {
    Exit,
    Help,
    Clear,
    New {
        name: String,
        address: String,
        number: String,
    },
    Search {
        search_string: String,
    },
    Delete {
        search_string: String,
    },
}

pub fn init_repl() {
    print!("{}[2J", 27 as char);
    print!("\x1B[2J\x1B[1;1H");
    println!("Welcome to your best phonebook application !");
    println!("Type 'help' to get a list of available commands.");
}

pub fn get_input() -> Input {
    loop {
        print!("> ");
        std::io::stdout().lock().flush().unwrap();
        let mut input = String::new();
        std::io::stdin().read_line(&mut input).unwrap();
        if input.is_empty() {
            return Input::Exit;
        }
        let mut input = input.split_whitespace();
        match input.next() {
            Some("exit") => return Input::Exit,
            Some("help") => return Input::Help,
            Some("clear") => return Input::Clear,
            Some("new") => {
                let name = input.next();
                let mut address: String = String::new();
                #[allow(unused_assignments)]
                let mut number: String = String::new();

                let mut tmp: String = String::new();
                loop {
                    let new = input.next();
                    if new.is_none() {
                        number = tmp;
                        break;
                    }
                    if !tmp.is_empty() {
                        address.push_str(&tmp.clone());
                        address.push(' ');
                    }
                    tmp = new.unwrap().to_string();
                }
                if name.is_none()
                    && address.is_empty()
                    && number.is_empty()
                    && name.unwrap().chars().nth(0).unwrap().is_numeric()
                {
                    println!("Invalid input");
                    continue;
                }
                return Input::New {
                    name: name.unwrap().to_string(),
                    address,
                    number,
                };
            }
            Some("search") => {
                let search_string = input.next();
                if search_string.is_none() {
                    println!("Invalid input");
                    continue;
                }
                return Input::Search {
                    search_string: search_string.unwrap().to_string(),
                };
            }
            Some("delete") => {
                let search_string = input.next();
                if search_string.is_none() {
                    println!("Invalid input");
                    continue;
                }
                return Input::Delete {
                    search_string: search_string.unwrap().to_string(),
                };
            }
            _ => {
                println!("Invalid input");
                continue;
            }
        };
    }
}

File: proc_phonebook/src/repl.cpp
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

#include "repl.hpp"

void init_repl()
{
    std::cout << "\x1B[2J\x1B[1;1H" << std::endl;
    std::cout << "Welcome to your best phonebook application !" << std::endl;
    std::cout << "Type 'help' to get a list of available commands."
              << std::endl;
}

std::vector<std::string> split(const std::string& s)
{
    std::vector<std::string> v;
    std::istringstream iss(s);
    std::string temp;
    while (iss >> temp)
        v.push_back(temp);
    return v;
}

Args get_input()
{
    std::string input;
    Args args = Args{.input = UNKNOWN, .datas = {}};
    std::cout << "> ";
    std::getline(std::cin, input);
    if (input.empty()) {
        args.input = EXIT;
        return args;
    }
    auto datas = split(input);

    auto it = datas.begin();

    std::string f = *it++;
    if (f == "exit")
        args.input = EXIT;
    else if (f == "help")
        args.input = HELP;
    else if (f == "clear")
        args.input = CLEAR;
    else if (f == "new") {
        args.input = ADD;
        std::string name = *it++;
        std::string address = "";
        std::string number = "";
        std::string tmp = "";
        while (true) {
            auto a = it++;
            if (a == datas.end()) {
                number = tmp;
                break;
            }
            if (!tmp.empty()) {
                address += tmp;
                address += " ";
            }
            tmp = *a;
        }
        if (name.empty() || address.empty() || number.empty()) {
            args.input = UNKNOWN;
            return args;
        }
        args.datas.push_back(name);
        args.datas.push_back(address);
        args.datas.push_back(number);
    } else if (f == "search") {
        auto str = *it++;
        if (str.empty()) {
            args.input = UNKNOWN;
            return args;
        }
        args.datas.push_back(str);
        args.input = SEARCH;
    } else if (f == "delete") {
        auto str = *it++;
        if (str.empty()) {
            args.input = UNKNOWN;
            return args;
        }
        args.datas.push_back(str);
        args.input = DELETE;
    } else {
        args.input = UNKNOWN;
    }
    return args;
}


